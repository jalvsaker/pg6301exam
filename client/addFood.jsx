import { useState } from "react";

export function AddFood({ reload }) {
  const [name, setName] = useState("");
  const [price, setPrice] = useState("");

  async function handleSubmit(e) {
    e.preventDefault();

    const body = { name, price };

    await fetch("/api/foods", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(body),
    });

    reload();
  }

  return (
    <form onSubmit={handleSubmit}>
      <label>
        Name:
        <input
          type={"text"}
          value={name}
          onChange={(e) => setName(e.target.value)}
        />
      </label>
      <label>
        Price:
        <input
          type={"number"}
          value={price}
          onChange={(e) => setPrice(e.target.value)}
        />
      </label>
      <button>Submit</button>
    </form>
  );
}
