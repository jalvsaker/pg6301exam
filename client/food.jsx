export function Food({ food }) {
  return (
    <div>
      {food.name} - {food.price}
      <div>
        <input type="number" /> <button>Add to cart</button>
      </div>
      <hr />
    </div>
  );
}
