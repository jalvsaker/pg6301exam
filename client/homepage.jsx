export function Homepage() {
  return <h1>Hello world</h1>;
}
