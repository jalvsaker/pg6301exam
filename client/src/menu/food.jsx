import React from "react";
import { useState } from "react";

export function Food({ food, setCart, disabled }) {
  const [amount, setAmount] = useState("0");

  function handleSubmit(e) {
    e.preventDefault();

    const newItem = { food, amount: parseInt(amount) };

    setCart((cart) => {
      return [
        ...cart.filter((item) => item.food._id !== newItem.food._id),
        newItem,
      ];
    });
  }

  return (
    <div>
      {food.name} - ${food.price}
      <form onSubmit={handleSubmit}>
        <label>
          Amount:
          <input
            type="number"
            value={amount}
            onChange={(e) => setAmount(e.target.value)}
            disabled={disabled}
          />{" "}
        </label>
        <button disabled={disabled}>Add to cart</button>
      </form>
      <hr />
    </div>
  );
}
