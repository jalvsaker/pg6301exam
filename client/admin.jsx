import { useNavigate } from "react-router-dom";
import { useEffect, useState } from "react";
import { AddFood } from "./addFood";

export function Admin({ user }) {
  const navigate = useNavigate();

  if (!user.isAdmin) {
    navigate("/");
  }

  const [foods, setFoods] = useState([]);
  const [loading, setLoading] = useState(true);
  const [reloader, setReloader] = useState(false);

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/foods");
      if (res.ok) {
        setFoods(await res.json());
        setLoading(false);
      }
    })();
  }, [reloader]);

  function reload() {
    setReloader((r) => !r);
  }

  if (loading) {
    return <div>Loading...</div>;
  }

  return (
    <>
      <h1>Admin panel</h1>

      <AddFood reload={reload} />

      {foods.map((food) => (
        <div key={food._id}>{food.name}</div>
      ))}
    </>
  );
}
