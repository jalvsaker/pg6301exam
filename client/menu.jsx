import { useEffect, useState } from "react";
import { Food } from "./food";

export function Menu() {
  const [foods, setFoods] = useState([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/foods");
      if (res.ok) {
        setFoods(await res.json());
        setLoading(false);
      }
    })();
  }, []);

  if (loading) {
    return <div>Loading...</div>;
  }

  return (
    <>
      <h1>Menu</h1>
      {foods.map((food) => (
        <Food food={food} key={food._id} />
      ))}
    </>
  );
}
