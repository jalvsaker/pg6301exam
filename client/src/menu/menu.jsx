import { useEffect, useState } from "react";
import { Food } from "./food";
import { Cart } from "./cart";
import { useNavigate } from "react-router-dom";

export function Menu({ user }) {
  const [foods, setFoods] = useState([]);
  const [loading, setLoading] = useState(true);

  const [cart, setCart] = useState([]);

  const navigate = useNavigate();

  async function order({ totalPrice, time }) {
    const res = await fetch("/api/orders", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ cart, totalPrice, time }),
    });

    if (res.ok) {
      navigate("/");
    }
  }

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/foods");
      if (res.ok) {
        setFoods(await res.json());
        setLoading(false);
      }
    })();
  }, []);

  if (loading) {
    return <div>Loading...</div>;
  }

  return (
    <>
      <h1>Menu</h1>
      {!user.username && <h4>Log in to order</h4>}
      {foods.map((food) => (
        <Food
          food={food}
          key={food._id}
          setCart={setCart}
          disabled={!user.username}
        />
      ))}
      <Cart cart={cart} order={order} disabled={!user.username} />
    </>
  );
}
