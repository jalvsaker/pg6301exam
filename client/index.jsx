import React from "react";
import { createRoot } from "react-dom/client";
import {
  BrowserRouter,
  Navigate,
  NavLink,
  Route,
  Routes,
} from "react-router-dom";
import { Homepage } from "./homepage";

function App() {
  return (
    <BrowserRouter>
      <nav>
        <NavLink to={"/"}>Home</NavLink>
        <NavLink to={"/test"}>Test</NavLink>
      </nav>
      <Routes>
        <Route path={"/"} element={<Homepage />} />
        <Route path={"/test"} element={<h2>test</h2>} />
        <Route path={"/*"} element={<Navigate to={"/"} />} />
      </Routes>
    </BrowserRouter>
  );
}

const root = createRoot(document.getElementById("root"));
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);
