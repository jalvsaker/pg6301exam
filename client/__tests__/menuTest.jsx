import { createRoot } from "react-dom/client";
import { act, Simulate } from "react-dom/test-utils";
import React from "react";
import { Menu } from "../src/menu/menu";
import { MemoryRouter } from "react-router-dom";
import { Food } from "../src/menu/food";

describe("menu test", function() {
  it("should show menu", async () => {
    const element = document.createElement("div");
    const root = createRoot(element);

    global.fetch = jest.fn(() => {
      return {
        ok: true,
        json: () => {
          return [];
        },
      };
    });

    await act(() => {
      root.render(
        <MemoryRouter>
        <Menu user={{username: "hei"}}/>
        </MemoryRouter>
      );
    });

    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should show Food", async () => {
    const element = document.createElement("div");
    const root = createRoot(element);

    const mock = jest.fn()

    await act(() => {
      root.render(
        <Food food={{name: "banana", price: 1}} setCart={mock}/>
      );
    });

    await act(() => {
      Simulate.change(element.querySelector("input"), {target: {value: "2"}})
    })

    await act(() => {
      Simulate.submit(element.querySelector("form"))
    })

    expect(element.innerHTML).toMatchSnapshot();
    expect(mock).toHaveBeenCalled()
  });
});