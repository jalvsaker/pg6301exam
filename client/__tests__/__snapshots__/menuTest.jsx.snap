// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`menu test should show Food 1`] = `"<div>banana - 1<form><label>Amount:<input type="number" value="2"> </label><button>Add to cart</button></form><hr></div>"`;

exports[`menu test should show menu 1`] = `"<h1>Menu</h1><h3>Cart:</h3><ul></ul><div>Price: 0.00</div><form><input type="time" value=""><button>Order</button></form><hr>"`;
